src/Api/PubnubApi/EventEngine/EventEngine.cs

```diff
@@ -0,0 +1,700 @@
+					LogCallback?.Invoke($"Received Message ({index + 1} of {receiveMessages.Length}) : {JsonConvert.SerializeObject(receiveMessages[index])}");
+					if (receiveMessages[index].Channel.IndexOf("-pnpres") > 0)
+					{
+						var presenceData = JsonConvert.DeserializeObject<PresenceEvent>(receiveMessages[index].Payload.ToString());
+					}
+					else
+					{
+						LogCallback?.Invoke($"Message : {JsonConvert.SerializeObject(receiveMessages[index].Payload)}");
+						PNMessageResult<object> messageResult = new PNMessageResult<object>();
+						messageResult.Channel = receiveMessages[index].Channel;
+						messageResult.Message = receiveMessages[index].Payload;
+						AnnounceMessage?.Invoke(messageResult);
+					}
+				}
+			}
+		}
+	}
+
+	public class HandshakeFailed : EffectInvocation
+	{
+		public List<string> Channels { get; set; }
+		public List<string> ChannelGroups { get; set; }
+	}
+	public class CancelHandshakeFailed : EffectInvocation
+	{
+
+	}
+    #endregion
+    public enum EventType
+	{
+		SubscriptionChanged,
+		SubscriptionRestored,
+		Handshake,
+		CancelHandshake,
+		HandshakeSuccess,
+		ReceiveMessages,
+		CancelReceiveMessages,
+		ReceiveSuccess,
+		HandshakeFailure,
+		CancelHandshakeFailure,
+		ReceiveFailure,
+		ReceiveReconnect,
+		CancelReceiveReconnect,
+		ReceiveReconnectSuccess,
+		ReceiveReconnectFailure,
+		ReceiveReconnectGiveUp,
+		HandshakeReconnect,
+		CancelHandshakeReconnect,
+		HandshakeReconnectSuccess,
+		HandshakeReconnectFailure,
+		HandshakeReconnectGiveUp,
+		ReconnectionFailed,
+		Disconnect,
+		Reconnect
+	}
+
+	public class ExtendedState
+	{
+		public List<string> Channels { get; set; }
+		public List<string> ChannelGroups { get; set; }
+		public long? Timetoken { get; set; }
+		public int? Region { get; set; }
+		public int Attempts { get; set; }
+
+		public ExtendedState()
+		{
+			Channels = new List<string>();
+			ChannelGroups = new List<string>();
+			Timetoken = 0;
+			Region = 0;
+			Attempts = 0;
+		}
+
+	}
+
+	public class EventEngine
+	{
+		private EventEngine pnEventEngine = null;
+
+		public ExtendedState Context;
+		public State? CurrentState { get; set; }
+		public List<State> States { get; set; }
+
+		public EffectDispatcher Dispatcher;
+
+		public EventEmitter Emitter;
+		public IPubnubUnitTest PubnubUnitTest { get; set; }
+
+		public EventEngine(EffectDispatcher dispatcher, EventEmitter emitter)
+		{
+			if (PubnubUnitTest != null )
+			{
+				PubnubUnitTest.EventTypeList?.Clear();
+			}
+			this.Dispatcher = dispatcher;
+			States = new List<State>();
+			Context = new ExtendedState();
+			this.Emitter = emitter;
+			emitter.RegisterHandler(this.Transition);
+		}
+
+		public State CreateState(StateType type)
+		{
+			var newState = new State(type);
+			if (States.Find(s=> s.StateType == type) != null) 
+			{
+				throw new InvalidOperationException($"StateType = {type} already exist.");
+			}
+			States.Add(newState);
+			return newState;
+		}
+
+		public void Transition(Event e)
```

Comment:
Since transtion will only do the job of transitioning state and handling over the invocation business to effect dispatcher. at high level it may look like this:

```c#
		public void Transition(Event e)
		{
			StateType nextStateType;
			if (CurrentState.transitions.TryGetValue(e.Type, out nextStateType)) {
				dispatcher.dispatch(CurrentState.ExitInvocationList);
				// update the context
				// Find the next state
				dispatcher.dispatch(NextState.EntryInvocationList);
				// update current state
				// dispatch managed effect
				if (CurrentState.Invocations.Count > 0) {	
						dispatcher.dispatch(CurrentState.Invocations, this.context);
					}
				}
			}
		}
```

Reply:
sure. we can add.

Reply:
While working on this refactor, I realized that we have abstract class and can reuse it.
Since we have abstract class `EffectInvocation` , I think we can add proposed members to the existing one, instead of creating interface.  What do you think?